### PR TITLE
Add more filter options to list_keys of S3Hook

### LINKED ISSUE #16627: add more filter options to list_keys of S3Hook
The hook has [list_keys](https://github.com/apache/airflow/blob/c8a628abf484f0bd9805f44dd37e284d2b5ee7db/airflow/providers/amazon/aws/hooks/s3.py#L265) allow to filter by prefix it would be nice to be able to filter by creation date of file or last modified date. ideally I think it would be nice if the function can support any kind of filter that boto3 allows.

The use case is that for the moment if you want to get all files that were modified after date X you need to list all the files and get them one by one to check their last modified date. this is not efficent.

### PR DESCRIPTION
**Pull request by sunank200**
Add more filter options to list_keys of S3Hook
    
This commit adds following filters to list the keys in list_keys of S3Hook:
    - `start_after_key` filters the any keys after the specified key. start_after_key can be any key in the bucket.
    - `start_after_datetime` filters all the keys with last modified date-time greater than or equal to the start_after_datetime.
    - `to_datetime` filters all the keys with last modified date-time less than or equal to the to_datetime
   
closes: #16627  

This change wouldn't affect dependencies for other operators like S3DeleteObjectsOperator, S3ListOperator, S3Hook methods:get_wildcard_key, delete_bucket and S3KeysUnchangedSensor.

Corresponding unittest has been added to test_s3.py.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by ferruzzi** [comment]
Is this functionality we want in the hooks?  Per the discussion [here](https://github.com/apache/airflow/pull/16571#discussion_r660433624) it sounds like that should be in the operator, maybe?

**Comment by kaxil** [comment]
Re-triggered CI

File: airflow/providers/amazon/aws/hooks/s3.py [hunk 1]
```diff
@@ -283,6 +287,12 @@ def list_keys(
         :type page_size: int
         :param max_items: maximum items to return
         :type max_items: int
+        :param start_after_key: returns keys after this specified key in the bucket.
+        :type start_after_key: str
+        :param start_after_datetime: returns keys with last modified datetime greater than the specified datetime.
+        :type start_after_datetime: datetime , ISO8601: '%Y-%m-%dT%H:%M:%S%z', e.g. 2021-02-20T05:20:20+0000
```
Comment (dstandish) [inline airflow/providers/amazon/aws/hooks/s3.py]: if the type is `datetime` then why do you need to specify string formatting here?  it makes it look like the type is actually string

File: airflow/providers/amazon/aws/hooks/s3.py [hunk 2]
```diff
@@ -269,6 +270,9 @@ def list_keys(
         delimiter: Optional[str] = None,
         page_size: Optional[int] = None,
         max_items: Optional[int] = None,
+        start_after_key: Optional[str] = '',
```
Comment (dstandish) [inline airflow/providers/amazon/aws/hooks/s3.py]: I think `None` is the standard default for optional string.  I think this is the most intuitive approach, compared with empty string.

Then you can you can add the parameter only conditionally to `paginate`.  To me it makes sense to only specify parameters when you actually want to use them.  `None` is a good default to mean, "do not use this param" (unless supplying `StartAfter=None` in paginate has the same effect as omitting it, in which case leaving it in is fine).

File: airflow/providers/amazon/aws/hooks/s3.py [hunk 3]
```diff
@@ -292,18 +302,24 @@ def list_keys(
...
         response = paginator.paginate(
-            Bucket=bucket_name, Prefix=prefix, Delimiter=delimiter, PaginationConfig=config
+            Bucket=bucket_name,
+            Prefix=prefix,
+            Delimiter=delimiter,
+            PaginationConfig=config,
+            StartAfter=start_after_key,
+        )
+        # JMESPath to query directly on paginated results
+        filtered_response = response.search(
```
Comment (dstandish) [inline airflow/providers/amazon/aws/hooks/s3.py]: why are you filtering even when `StartAfter` is `None`?

File: airflow/providers/amazon/aws/hooks/s3.py [hunk 4]
```diff
@@ -292,18 +302,24 @@ def list_keys(
...
+            "Contents[?to_string("
+            "LastModified)<='\"{}\"' && "
+            "to_string(LastModified)>='\"{"
```
Comment (dstandish) [inline airflow/providers/amazon/aws/hooks/s3.py]: if it's going to be called `start_after_datetime` then this should be strictly greater than, i think

**Review by dstandish — CHANGES_REQUESTED** [review]
Let me offer a recommendation.

Rather than adding `start_after_datetime` and `to_datetime`, just add a single param called `object_filter` that takes a function `(obj: dict) -> bool`

then this function can be applied like so:

```python
        for page in response:
            if 'Contents' in page:
                for k in page['Contents']:
                    <evaluate object filter here and only append contditionally>
                    keys.append(k['Key'])
```

with obj filter people can apply whatevery logic they want and they don't have to understand what start_from_datetime / to_datetime mean and how to use (since they aren't part of the s3 api)

what do yall think

**Comment by dstandish** [comment]
> Let me offer a recommendation.
> 
> Rather than adding `start_after_datetime` and `to_datetime`, just add a single param called `object_filter` that takes a function `(obj: dict) -> bool`
> 
> then this function can be applied like so:
> 
> ```python
>         for page in response:
>             if 'Contents' in page:
>                 for k in page['Contents']:
>                     <evaluate object filter here and only append contditionally>
>                     keys.append(k['Key'])
> ```
> 
> with obj filter people can apply whatevery logic they want and they don't have to understand what start_from_datetime / to_datetime mean and how to use (since they aren't part of the s3 api)
> 
> what do yall think

Maybe this isn't the greatest idea...  I thought there were more attributes available in the response.

Here's what you get in the response:

```
{'Key': 'tmp/hello/1.txt',
 'LastModified': datetime.datetime(2021, 7, 6, 3, 13, 41, tzinfo=tzutc()),
 'ETag': '"49f68a5c8493ec2c0bf489821c21fc3b"',
 'Size': 2,
 'StorageClass': 'STANDARD'}
```

So at least for now there isn't much else in that response that you'd want to filter on.

Still, putting the filtering in the user's hands would remove the ambiguity of strictly greater than vs greater than (and same with less than)?  

Also there's the problem of naming.

If we include both parameters `from` and `to`, we should probably call them `from_datetime` and `to_datetime`, where `from` is `>=` and `to` is `<`.  

One other thing.... I am not sure why we use the jmespath when we already have the datetime object available in the for loop in the existing code.  I think it is best to compare `LastModified` directly with the user supplied datetime rather than converting everything to string and doing a json search.  For one this is additional transformation that inherently could introduce an error through formatting difference.  I also suspect the json search is significantly less efficient but not sure.  Bottom line, if you already have datetime on both sides of the comparison, why convert them both to string to compare them?

File: airflow/providers/amazon/aws/hooks/s3.py (hunk 4 above)
Comment (dstandish) [reply airflow/providers/amazon/aws/hooks/s3.py]: but i recommend just `from_datetime` and `to_datetime`

Comment (dstandish) [inline airflow/providers/amazon/aws/hooks/s3.py]: does not handle case where you have exactly _one of_ `to_datetime` and `start_after_datetime` (i.e. not both)

also pendulum and datetime are formatted differently. observe:
```python
>>> import pendulum
>>> '{}'.format(pendulum.now())
'2021-07-06T21:22:11.514220-07:00'
>>> from datetime import datetime
>>> '{}'.format(datetime.now())
'2021-07-06 21:22:35.418199'
```

if you compare the objects directly, your feature could support both pendulum _(which is used commonly with airflow)_ and datetime; but by using string formatting you really cannot use both.

File: airflow/providers/amazon/aws/hooks/s3.py (hunk 2 above)
Comment (pateash) [reply airflow/providers/amazon/aws/hooks/s3.py]: Completely agree with @dstandish,
I have rarely seen a default string as literal '' in my experience in Airflow codebase.

Comment (pateash) [reply airflow/providers/amazon/aws/hooks/s3.py]:
```suggestion
        start_after_key: Optional[str] = None,
```

**Review by pateash — CHANGES_REQUESTED** [review]

**Comment by potiuk** [comment]
Hey @sunank200  - how about applying the comments from reviews ? I am preparing to release next wave of providers and it would be great to merge that one in.

**Comment by sunank200** [comment]
@potiuk sure. Was planning to comment on the review and do changes today. Some situation at home hence the delay. I will resolve the issues today.

**Comment by potiuk** [comment]
> @potiuk sure. Was planning to comment on the review and do changes today. Some situation at home hence the delay. I will resolve the issues today.

No worries. Hope all is good !

File: airflow/providers/amazon/aws/hooks/s3.py (hunk 3 above)
Comment (sunank200) [reply airflow/providers/amazon/aws/hooks/s3.py]: When `StartAfter` is `None`, `paginator.paginate()` doesn't use it as filter. It filters on `StartAfter` only when it's not a `None`. So, filtering works in both the conditions when it's `None` and not `None`.  But like you have suggested above, I will add a condition to using the parameters only when passed by the user as argument.

File: airflow/providers/amazon/aws/hooks/s3.py (hunk 4 above)
Comment (sunank200) [reply airflow/providers/amazon/aws/hooks/s3.py]: sure. `from_datetime` and `to_datetime` make more sense. I will do the change.

File: airflow/providers/amazon/aws/hooks/s3.py (hunk 1 above)
Comment (sunank200) [reply airflow/providers/amazon/aws/hooks/s3.py]: @dstandish I don't think we require to specify the string formatting here. `datetime` object should be passed here as an argument, it will work fine. String type passed as an argument won't work. I was just trying to give example. I will update the examples as:
```
e.g:  to_datetime =datetime.strptime('2024-08-19T09:55:48+0000','%Y-%m-%dT%H:%M:%S%z')
```

File: airflow/providers/amazon/aws/hooks/s3.py (hunk 2 above)
Comment (sunank200) [reply airflow/providers/amazon/aws/hooks/s3.py]: This makes sense. I will add the condition to use `StartAfter` only when it's not `None`.  Also, I will make the default string `None` as per the Airflow codebase.

**Review by george79d — APPROVED** [review]

**Review by george79d — APPROVED** [review]
Hello Cyprus

File: airflow/providers/amazon/aws/hooks/s3.py (hunk 4 above)
Comment (sunank200) [reply airflow/providers/amazon/aws/hooks/s3.py]: @potiuk  @dstandish @pateash I have pushed the changes just now.

Created `ResponseFilter` class to parse and filter the s3 boto3 response based on various operations defined by the user.  This class can be extended as required for parsing and filtering the response based on different object filters operation. For now, I have added the following allowed operations.
```
 allowed_operations = {
        "lte": operator.le,
        "gte": operator.ge,
        "gt": operator.gt,
        "lt": operator.lt,
        "eq": operator.eq,
    }
```

`ResponseFilter.filter(object_filter=object_filter)`  puts the filtering in the user's hands (through the use of a filter and would remove the ambiguity of strictly greater than vs greater than (and same with less than). Please refer to the commits in this [thread](https://github.com/apache/airflow/pull/16821/commits/4ef750ea486f76c98f7e5927b12fc62b4767913c) and [test changes](https://github.com/apache/airflow/pull/16821/commits/a7468c0d1591c83fd1a1c7f3ca8e32a2750e4f5b).

I have added the comment for the implementation details in the code.

**Comment by sunank200** [comment]
@potiuk  @dstandish @pateash I have done the required changes. I would be more than happy to accommodate the feedback.

File: airflow/providers/amazon/aws/hooks/s3.py (hunk 2 above)
Comment (sunank200) [reply airflow/providers/amazon/aws/hooks/s3.py]: Done with the changes

File: airflow/providers/amazon/aws/hooks/s3.py (hunk 4 above)
Comment (sunank200) [reply airflow/providers/amazon/aws/hooks/s3.py]: I have removed `from_datetime`, `to_datetime` and using `ResponseFilter` class with the filter method which supports all the operations defined by user: [commit](https://github.com/apache/airflow/pull/16821/commits/4ef750ea486f76c98f7e5927b12fc62b4767913c)

Comment (dstandish) [reply airflow/providers/amazon/aws/hooks/s3.py]: i'm not sure adding this ResponseFilter is the right approach.  i am not sure we need to add this abstraction layer.

Why not just let the user put an arbitrary callable that does the filtering that the user needs?  

For example, your docstring shows this example:

```python
object_filter={"LastModified__lt": datetime.now()},
```

But with arbitrary callable, this can be implemented about as simply this way:

```python
object_filter=lambda x: x['LastModified'] < datetime.now(),
```

This is roughly as compact and roughly as simple, but the benefit is there's no need to understand the options available within the class and how it works, and if user wants to do arbitrary filtering they can.

---

That said, I still think it's probably best to go with your original thought, i.e. `from_datetime` and `to_datetime` (though with the updated naming / logic).  

_(I know I initially suggested adding object filter callable, but subsequently I walked back that suggestion; since there aren't other useful filter attributes in the object metadata, it's not as useful)_.

And your approach of adding only a certain number of allowable operations doesn't really provide the flexibility of the callable.

Simply adding the two optional datetime params, from and to, is simple, easy to understand, useful.  It's true that with simple from / to params, we can't provide flexibility with interval edge inclusivity, but to me that seems like an acceptable tradeoff for the simplicity and usabilitiy.  If we go with this approach I would probably vote for doing `>=` for `from` and `<` for `to`.

---

In conclusion i think going with `from` and `to` seems best, but if not that then simple arbitrary callable instead of the ResponseFilter apparatus seems best.  Interested in what others think.

**Review by pateash — APPROVED** [review]

File: airflow/providers/amazon/aws/hooks/s3.py (hunk 4 above)
Comment (sunank200) [reply airflow/providers/amazon/aws/hooks/s3.py]: Sure @dstandish, I think the `to` and `from` approach keeps the code simple and doesn't add additional complexity to it. `ResponseFilter` and `object_filter` callable in my opinion makes things more generic for any filter operations in the future but it does add additional complexity to it. Would be happy to accommodate the change once others comment and conclude on the approach.

@potiuk what do you think?

**Comment by potiuk** [comment]
Can you please rebase to latest main @sunank200 ? We had some failure in main recently and they are all fixed now.

File: airflow/providers/amazon/aws/hooks/s3.py (hunk 4 above)
Comment (sunank200) [reply airflow/providers/amazon/aws/hooks/s3.py]: @dstandish I thought about it, if we take `ResponseFilter` route, it will add additional complexity to the code and people might need to understand the approach. Hence for simplicity of the codebase, I am going forward with `from` and `to` approach. I would be renaming it to `lastmodified_datetime_gte` (>=) and `lastmodified_datetime_lt` (<) .

**Comment by sunank200** [comment]
@potiuk I have rebase to latest main.

**Comment by potiuk** [comment]
Reopened it to re-trigger the build

**Comment by potiuk** [comment]
There are some static checks and docs build that needs fixes (I recommend installing pre-commit to fix the static checks)

**Comment by sunank200** [comment]
Have implemented the pre-commit hook changes.
